monomerdb/localdb/localdb.go

```diff
@@ -0,0 +1,464 @@
+)
+
+var (
+	endian            = binary.BigEndian
+	unsafeLabelKey    = bucketHashByLabel.Key([]byte(eth.Unsafe))
+	safeLabelKey      = bucketHashByLabel.Key([]byte(eth.Safe))
+	finalizedLabelKey = bucketHashByLabel.Key([]byte(eth.Finalized))
+	heightKey         = bucketHeight.Key()
+)
+
+type DB struct {
+	db *pebble.DB
+}
+
+func New(db *pebble.DB) *DB {
+	return &DB{
+		db: db,
+	}
+}
+
+// TODO: optimization - we can use a cbor encoder to write directly to a batch, rather than copying bytes twice.
+
+// AppendBlock does no validity checks and does not update labels.
+func (db *DB) AppendBlock(block *monomer.Block) error {
+	headerBytes, err := cbor.Marshal(block.Header)
+	if err != nil {
+		return fmt.Errorf("marshal header into cbor: %v", err)
+	}
+	heightBytes := marshalUint64(uint64(block.Header.Height))
+	return db.update(func(b *pebble.Batch) error {
+		if err := b.Set(bucketHeaderByHeight.Key(heightBytes), headerBytes, nil); err != nil {
+			return fmt.Errorf("set block by height: %v", err)
+		}
+		if err := b.Set(bucketHeightByHash.Key(block.Header.Hash.Bytes()), heightBytes, nil); err != nil {
+			return fmt.Errorf("set height by hash: %v", err)
+		}
+		if err := b.Set(heightKey, heightBytes, nil); err != nil {
+			return fmt.Errorf("set height: %v", err)
+		}
+
+		for i, tx := range block.Txs {
+			heightAndIndexBytes := slices.Concat(heightBytes, marshalUint64(uint64(i)))
+			if err := b.Set(bucketTxByHeightAndIndex.Key(heightAndIndexBytes), tx, nil); err != nil {
+				return fmt.Errorf("set tx by height and index: %v", err)
+			}
+			if err := b.Set(bucketTxHeightAndIndexByHash.Key(tx.Hash()), heightAndIndexBytes, nil); err != nil {
+				return fmt.Errorf("set tx height and index by hash: %v", err)
+			}
+		}
+		return nil
+	})
+}
+
+func (db *DB) UpdateLabels(unsafe, safe, finalized common.Hash) error {
+	return db.update(func(b *pebble.Batch) error {
+		return updateLabels(b, unsafe, safe, finalized)
+	})
+}
+
+func updateLabels(b *pebble.Batch, unsafe, safe, finalized common.Hash) error {
+	if err := b.Set(unsafeLabelKey, unsafe.Bytes(), nil); err != nil {
+		return fmt.Errorf("set unsafe hash by label: %v", err)
+	}
+	if err := b.Set(safeLabelKey, safe.Bytes(), nil); err != nil {
+		return fmt.Errorf("set safe hash by label: %v", err)
+	}
+	if err := b.Set(finalizedLabelKey, finalized.Bytes(), nil); err != nil {
+		return fmt.Errorf("set finalized hash by label: %v", err)
+	}
+	return nil
+}
+
+// Rollback rolls back the chain and updates labels.
+func (db *DB) Rollback(unsafe, safe, finalized common.Hash) error {
+	return db.updateIndexed(func(b *pebble.Batch) (err error) {
+		unsafeHeightBytesValue, closer, err := get(b, bucketHeightByHash.Key(unsafe.Bytes()))
+		if err != nil {
+			return fmt.Errorf("get height by hash %s: %w", unsafe, err)
+		}
+		defer func() {
+			err = wrapCloseErr(err, closer)
+		}()
+		unsafeHeight := endian.Uint64(unsafeHeightBytesValue)
+		firstHeightBytesToDelete := marshalUint64(unsafeHeight + 1)
```

Comment:
curious - why do we rollback to the unsafe height here instead of the safe height?

Reply:
There are always three heights, unsafe, safe, and finalized. 

unsafe >= safe >= finalized

If we roll back to safe, we might delete unsafe blocks.